src/game/client/draw/w3dmodeldraw.cpp

```diff
@@ -1154,3 +1178,1890 @@ void W3DModelDrawModuleData::Parse_Condition_State(INI *ini, void *instance, voi
+
+    Drawable *drawable = Get_Drawable();
+
+    if (drawable != nullptr) {
+        Object *object = drawable->Get_Object();
+
+        if (object != nullptr) {
+            if (g_theWriteableGlobalData->m_timeOfDay == TIME_OF_DAY_NIGHT) {
+                m_hexColor = object->Get_Night_Indicator_Color();
+            } else {
+                m_hexColor = object->Get_Indicator_Color();
+            }
+        }
+
+        if (!Get_W3D_Model_Draw_Module_Data()->m_recievesDynamicLights) {
+            drawable->Set_Recieves_Dynamic_Lights(false);
+            captainslog_debug("Set_Recieves_Dynamic_Lights = FALSE: %s", drawable->Get_Template()->Get_Name().Str());
+        }
+    }
+
+    Set_Model_State(idle);
+}
+
+W3DModelDraw::~W3DModelDraw()
+{
+    if (m_trackRenderObject != nullptr && g_theTerrainTracksRenderObjClassSystem != nullptr) {
+        g_theTerrainTracksRenderObjClassSystem->Unbind_Track(m_trackRenderObject);
+        m_trackRenderObject = nullptr;
+    }
+
+    Nuke_Current_Render(nullptr);
+}
+
+void W3DModelDraw::On_Drawable_Bound_To_Object()
+{
+    Get_W3D_Model_Draw_Module_Data()->Validate_Stuff_For_Time_And_Weather(Get_Drawable(),
+        g_theWriteableGlobalData->m_timeOfDay == TIME_OF_DAY_NIGHT,
+        g_theWriteableGlobalData->m_weather == WEATHER_SNOWY);
+}
+
+void W3DModelDraw::Do_Start_Or_Stop_Particle_Sys()
+{
+    Drawable *drawable = Get_Drawable();
+    bool hidden = drawable->Is_Hidden() || m_fullyObscuredByShroud;
+
+    for (auto &tracker : m_particleSystemIDs) {
+        ParticleSystem *psys = g_theParticleSystemManager->Find_Particle_System(tracker.id);
+
+        if (psys != nullptr) {
+            if (hidden) {
+                psys->Stop();
+            } else {
+                psys->Start();
+            }
+        }
+    }
+}
+
+void W3DModelDraw::Set_Hidden(bool hidden)
+{
+    if (m_renderObject != nullptr) {
+        m_renderObject->Set_Hidden(hidden);
+    }
+
+    if (m_shadow != nullptr) {
+        m_shadow->Enable_Shadow_Render(!hidden);
+    }
+
+    m_hidden = hidden;
+
+    if (m_decalShadow != nullptr) {
+        m_decalShadow->Enable_Shadow_Render(!hidden);
+    }
+
+    if (m_trackRenderObject != nullptr && hidden) {
+        const Coord3D *pos = Get_Drawable()->Get_Position();
+        m_trackRenderObject->Add_Cap_Edge_To_Track(pos->x, pos->y);
+    }
+
+    Do_Start_Or_Stop_Particle_Sys();
+}
+
+void W3DModelDraw::Release_Shadows()
+{
+    if (m_shadow != nullptr) {
+        m_shadow->Release();
+        m_shadow = nullptr;
+    }
+}
+
+void W3DModelDraw::Allocate_Shadows()
+{
+    Shadow::ShadowTypeInfo info;
+    info.m_thing = Get_Drawable()->Get_Template();
+
+    if (m_shadow == nullptr && m_renderObject != nullptr && g_theW3DShadowManager != nullptr
+        && info.m_thing->Get_Shadow_Type() != SHADOW_NONE) {
+        strcpy(info.m_shadowName, info.m_thing->Get_Shadow_Texture_Name());

```

Comment:
break these apart so its not hard to look at

Reply:
Added some blank lines.